# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/Marking.cpp
# Commit: 967fd2ce6f67
# Full Hash: 967fd2ce6f67b87426c0878cac53d461165ca1f1
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-10-29 09:50:10
# Regressor Bug: 1797755
# File Overlap Count: 1
# Description:
#   Bug 1797755 - Part 2: Make delayed marking colors work like normal marking r=sfink
#   
#   Currently the way mark colors work for delayed marking doesn't align with
#   normal marking for gray marking of GC things that can only be marked black
#   (e.g. strings).
# ==============================================================================

diff -r cf090cf39fad -r 967fd2ce6f67 js/src/gc/Marking.cpp
--- a/js/src/gc/Marking.cpp	Fri Oct 28 15:17:42 2022 +0000
+++ b/js/src/gc/Marking.cpp	Fri Oct 28 15:17:42 2022 +0000
@@ -2009,21 +2009,20 @@
     markLaterArenas++;
 #endif
   }
-  JS::TraceKind kind = MapAllocToTraceKind(arena->getAllocKind());
-  MarkColor colorToMark =
-      TraceKindCanBeMarkedGray(kind) ? markColor() : MarkColor::Black;
-  if (!arena->hasDelayedMarking(colorToMark)) {
-    arena->setHasDelayedMarking(colorToMark, true);
+
+  if (!arena->hasDelayedMarking(markColor())) {
+    arena->setHasDelayedMarking(markColor(), true);
     delayedMarkingWorkAdded = true;
   }
 }
 
 void GCMarker::markDelayedChildren(Arena* arena) {
   JS::TraceKind kind = MapAllocToTraceKind(arena->getAllocKind());
-  MOZ_ASSERT_IF(markColor() == MarkColor::Gray, TraceKindCanBeMarkedGray(kind));
+  MarkColor colorToCheck =
+      TraceKindCanBeMarkedGray(kind) ? markColor() : MarkColor::Black;
 
   for (ArenaCellIterUnderGC cell(arena); !cell.done(); cell.next()) {
-    if (cell->isMarked(markColor())) {
+    if (cell->isMarked(colorToCheck)) {
       JS::TraceChildren(this, JS::GCCellPtr(cell, kind));
     }
   }
